crates/typst-library/src/text/quotes.rs

```diff
@@ -228,14 +277,117 @@ impl<'s> Quotes<'s> {
     }
 }
 
-impl Default for Quotes<'_> {
-    /// Returns the english quotes as default.
-    fn default() -> Self {
-        Self {
-            single_open: "‘",
-            single_close: "’",
-            double_open: "“",
-            double_close: "”",
-        }
+/// An opening and closing quote.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct QuoteSet {
+    open: EcoString,
+    close: EcoString,
+}
+
+cast! {
+    QuoteSet,
+    self => {
+        [self.open, self.close]
+            .into_iter()
+            .map(IntoValue::into_value)
+            .collect::<Array>()
+            .into_value()
+    },
+    value: Array => {
+        let [open, close] = array_to_set(value)?;
+        Self { open, close }
+    },
+    value: Str => {
+        let [open, close] = str_to_set(value.as_str())?;
+        Self { open, close }
+    },
+}
+
+fn str_to_set(value: &str) -> StrResult<[EcoString; 2]> {
+    let mut iter = value.graphemes(true);
+    let open = iter
+        .next()
+        .ok_or_else(|| eco_format!("expected 2 characters, got 0 characters"))?;
+    let close = iter
+        .next()
+        .ok_or_else(|| eco_format!("expected 2 characters, got 1 character"))?;
+
+    let count = iter.count();
+    if count > 0 {
+        bail!(
+            "expected 2 characters, got {count} character{}",
+            if count > 1 { "s" } else { "" }
+        );
     }
+
+    Ok([open, close].map(EcoString::from))
+}
+
+fn array_to_set(value: Array) -> StrResult<[EcoString; 2]> {
+    let value = value.as_slice();
+    if value.len() != 2 {
+        bail!(
+            "expected 2 quotes, got {} quote{}",
+            value.len(),
+            if value.len() > 1 { "s" } else { "" }
+        );
+    }
+
+    let open: EcoString = value[0].clone().cast()?;
+    let close: EcoString = value[1].clone().cast()?;
+
+    Ok([open, close])
+}
+
+/// A dict of single and double quotes.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct QuoteDict {
+    double: Smart<QuoteSet>,
+    single: Smart<QuoteSet>,
+}
+
+fn map_smart(value: Value) -> StrResult<Smart<QuoteSet>> {
+    Ok(match value {
+        Value::Auto => Smart::Auto,
+        v => Smart::Custom(QuoteSet::from_value(v)?),
+    })
+}
+
+cast! {
+    QuoteDict,
+    self => {
+        [("double", self.double), ("single", self.single)]
+            .into_iter()
+            .map(|(k, v)| (k.into(), v.into_value()))
+            .collect::<Dict>()
+            .into_value()
+    },
+    mut value: Dict => {
+        let keys = ["double", "single"];
+
+        let double = value
+            .take("double")
+            .ok()
+            .map(map_smart)
```

Comment:
```suggestion
            .map(FromValue::from_value)
```
same below. then we don't need `map_smart`